### ISSUE TITLE
Significant input lag from the FBE

### ISSUE THREAD

**Issue by false-9** [gonetz/GLideN64#1822]
mupen64plus-gui build May 18 2018
Plugin: GlideN64

When disabling FBEmulation, input lag is reduced significantly in games. It emphasises how unresponsive some games are with FBE enabled, Ocarina of Time for example has roughly 10-11 frames of input latency with FBE on, and around 6-7 frames with it off. This is measured while excluding display, peripheral or any other external lag.

Likewise, on Mario 64, I measured ~6 frames of latency with FBE on and 0 latency with FBE off. Can the FBE be improved to reduce input latency?

Specs: i7 3770k, Nvidia GTX 670, 16GB RAM

**Comment by false-9** [comment]
While I do think disablng FBE might be going negative (- lag count) compared to original hardware's behaviour, I do think with it on it is definitely over it by a margin. As loganmc10 it does depend on the game, some games it has no effect, take No Mercy as an example, with FBE on/off it doesn't matter, the frame count before input displayed is the same. So I do think there's something odd at play, and something to be rectified.

I don't have the luxury of a high speed camera, or a LED mapped to button input controller to measure native behaviour frame for frame, but I can say with confidence there's a big difference just by feel, it's noticeable instantly, granted display lag is removed with the CRT, my monitor is low latency however, so there's no way it's just display lag.

**Comment by loganmc10** [comment]
> take No Mercy as an example, with FBE on/off it doesn't matter, the frame count before input displayed is the same. So I do think there's something odd at play, and something to be rectified.

I think the fact that it varies by game should show you that it is not the emulator or video plugin at fault. The video plugin doesn't have game specific latency settings. The fact that the video plugin is capable of having low input lag on some games (No Mercy, Smash Bros) shows me that it is working properly.

Modern computes have some things that are going to add a bit of latency, like double buffering, LCDs have more latency than CRTs, etc...

Either way, it's not worth someone spending time on it unless someone can actually prove there is something inaccurate about how the emulator/video plugin works compared to a real N64

**Comment by false-9** [comment]
Fair enough.

I'm not fully convinced but I'll take your word for it until I can measure it accurately between systems. Although going by anecdotes, Mario 64 reportedly has a 3 frame latency on original hardware, so judging by the ~6 frames at play on GlideN with FBE via RA's frame advance makes me dubious that it's entirely accurate.

**Comment by false-9** [comment]
Then perhaps you are right and disabling FBE is merely compensating for the other sources of latency so that it feels closer to original hardware on a CRT. I just find it difficult to believe OOT as one example has that many inherent frames of lag and it certianly feels sluggish to play. I'll try to properly measure it whenever i get the chance and report back. Close the issue if you want.

Cheers.

**Comment by theboy181** [comment]
It has been my experience to trust your gut. People still tell me they can’t see the dropped frames in ssb64.

**Comment by false-9** [comment]
Yeah, it's frustrating because it's a hard thing to discuss with people and difficult to quantify, but I've always had a good "radar" so to speak when things seem off in a certain way, I'm like this with modern PC games as well.

**Comment by LegendOfDragoon** [comment]
>People still tell me they can’t see the dropped frames in ssb64

Maybe because its not as laggy for them :smile:

**Comment by theboy181** [comment]
Your eyes be broken son. And since you are using my old laptop I know that it’s your eyes and not the superior HW.

Sent from my iPhone

On May 24, 2018, at 12:22 PM, LegendOfDragoon <notifications@github.com<mailto:notifications@github.com>> wrote:


People still tell me they can’t see the dropped frames in ssb64

Maybe because its not as laggy for them 😄

**Comment by LegendOfDragoon** [comment]
>And since you are using my old laptop I know that it’s your eyes and not the superior HW.

I never even stated which hardware I've done extensive testing on, for frame rate/ input latency :smile: . Anyway i managed to improve it somewhat on that laptop by tweaking settings today (still not at ideal smoothness).

@false-9 what OS are you using and are you using vsync? With vsync off, no FBE definitely feels smoother and less delay. Harder to tell, with vsync on, on my end. Aero is absolutely disgusting and I can't stand it at all.. Best you can do is full screen exclusive mode.

**Comment by false-9** [comment]
@LegendOfDragoon Windows 10. Yeah I'm always using exclusive fullscreen if the options there. Vsync off definitely helps matters but that's quite usual for most applications and is outside the cause with FBE.

Mario Kart 64, OoT, and Mario 64 are just 3 examples of games that all show significant and very noticeable differences of latency reduction with FBE off, but the question that remains is what is the accurate amount of frames that those games should have on an actual N64, but it's certain that FBE off, for better or for worse significantly reduces latency with cause and effect. My suspicion is that some games don't play nice with the implementation of it, while others are fine, because even if No Mercy and Smash show no difference, other games most definitely do, so why is that happening? If no FBE breaks something which causes super responsiveness in _some_ games, why not all of the games? That's what I don't get, and that's why I have my doubts.

**Comment by loganmc10** [comment]
> If no FBE breaks something which causes super responsiveness in some games, why not all of the games?

It's a little beyond the scope of this discussion, but it's because with FBE off, the video plugin ignores the VI_ORIGIN register. Basically some games (like Mario 64) are obviously implementing their own double/triple buffering. They create a picture, but they don't immediately send that picture to the monitor. With FBE off the video plugin ignores these buffering systems. Some games buffer more images than others

You can test the Angrylion video plugin and you'll observe the exact same behaviour as GLideN64 with FBE enabled

**Comment by false-9** [comment]
Ahh, I see, that makes sense. TIL.

Then I guess I'll have to concede. OoT must just be really sluggish by design with it's low framerate etc. but you don't notice it as much on N64/CRT without all the external factors that come into it on PC like you talked about.

I'm interested now though, and forgive my ignorance, is it technically possible to have an FBE that ignores the VI_ORIGIN register but outputs images correctly as it does now? In some games FBE off obviously introduces some artificating here and there in certain games, I know it's heresey to suggest an inaccuracy as a feature like that but honestly, with the gains, if done correctly it could seriously compensate for some of the inherent lag by Vsync, Controllers etc. while most of my enjoyment, and lack thereof with emulator use stems from input latency.

**Comment by loganmc10** [comment]
> is it technically possible to have an FBE that ignores the VI_ORIGIN register but outputs images correctly as it does now?

@gonetz could answer much better, but I think that's kind of like asking "can you do framebuffer emulation without the framebuffers?"

**Comment by false-9** [comment]
Yeah I thought it might come across like that, haha! I just remember reading about how the Dolphin team achieved something similar with that though, i.e. less latency than the system with one of the buffer options.

**Comment by theboy181** [comment]
Set vsync and turn off framelimit for perfect FPS. :) tested on that laptop myself. :)

**Comment by LegendOfDragoon** [comment]
Turning off frame limiter does help, when vsync is enabled. But after doing more testing, I have come to the conclusion that Windows 10 is bad for gaming. Aero is atrocious and seems to even affect fullscreen. I will never do any serious gaming on Windows 10.

Anyway @false-9 do you mind testing Glide64? I seem to get less input lag with that plugin, than with GLideN64 + FBE enabled.

**Comment by theboy181** [comment]
This issue was the same in WIN7 on that same laptop.. I think your conclusion on Windows 10 is unlrealted to the dropped frames.

**Comment by false-9** [comment]
@LegendOfDragoon Haven't had the time until recently, to answer your question I installed latest GlideN64 on PJ64 and then tested it side by side with Glide64, simply because I found trying to install Glide64 on Mupen64Plus a pain in the ass. So anyway yeah, Glide64 is instantly noticeably more responsive than GlideN64, both with their respective FBE's turned on.
